Various high DPI fixes

drewnoakes opened the pull request:
Relates to #4775.

Changes proposed in this pull request:
- Explicitly scale `DataGridViewColumn.Width` for high DPI support
- Fix button icon sizes on `FormCommit`
- Fix list item icon sizes on `FileStatusList`
- Remove some unused code

What did I do to test the code and ensure quality:
- Lots of manual testing at different scaling factors.

Has been tested on:
 - Windows 10

---

## Screenshots (before / after)

### Revision grid

- Fixes sizing of nodes in graph.
- Fixes width of columns in grid. Especially noticeable for multi-line indicator and date/time.

Before

![image](https://user-images.githubusercontent.com/350947/38310903-a2b095bc-3815-11e8-813d-322947841ba6.png)

After

![image](https://user-images.githubusercontent.com/350947/38310920-ae63212c-3815-11e8-94ad-4088c18410a9.png)

---

### File Tree

Before

![image](https://user-images.githubusercontent.com/350947/38341815-6c5394ee-3872-11e8-94e0-12df3f88ee57.png)

After 

![image](https://user-images.githubusercontent.com/350947/38341825-784d2a8a-3872-11e8-8674-ebbb731e3ec5.png)


---

### FormCommit

Before

![image](https://user-images.githubusercontent.com/350947/38341730-f00850be-3871-11e8-8cb9-9f6dcd0f3382.png)

After

![image](https://user-images.githubusercontent.com/350947/38341715-df956b36-3871-11e8-973c-c132b1ac4cf9.png)

---

### Patch form

Before

![image](https://user-images.githubusercontent.com/350947/38311038-0127c99e-3816-11e8-8709-3b49a67163bc.png)

After

![image](https://user-images.githubusercontent.com/350947/38311052-105cd9c2-3816-11e8-9dfd-ad95cea9b269.png)

---

### Dashboard settings

Before

![image](https://user-images.githubusercontent.com/350947/38311085-2ef9d59c-3816-11e8-9966-c7555a9db95a.png)

After

![image](https://user-images.githubusercontent.com/350947/38311096-3ce853e0-3816-11e8-9e05-766dd0771615.png)

---

### Revision links

Before

![image](https://user-images.githubusercontent.com/350947/38311832-02e8ead6-3818-11e8-80aa-06e5c9939b63.png)

After

![image](https://user-images.githubusercontent.com/350947/38311845-0cbc9256-3818-11e8-8a76-ed2f4ba1400e.png)


---

### Verify form

Before

![image](https://user-images.githubusercontent.com/350947/38311584-720a59dc-3817-11e8-8f39-7aeecef18402.png)

After

![image](https://user-images.githubusercontent.com/350947/38311621-8d6aa164-3817-11e8-8e0f-8622ba43ceb9.png)

---

### View patch

Before

![image](https://user-images.githubusercontent.com/350947/38311301-bed201b2-3816-11e8-9c7e-d655d72b6ef9.png)

After

![image](https://user-images.githubusercontent.com/350947/38314682-b68714ea-381e-11e8-949b-d7bfa9734460.png)

---

### Delete unused branches

Before

![image](https://user-images.githubusercontent.com/350947/38311226-92cf5f74-3816-11e8-87ed-3a71e3cdf29c.png)

After

![image](https://user-images.githubusercontent.com/350947/38311233-99a4e88c-3816-11e8-8e86-98bcb71bab7f.png)

drewnoakes commented:
There appears to be a regression in the "view patch" example. I'll investigate.

RussKie commented:
Could you please also create a MVP for 2.51 branch

RussKie commented:
Please submit the fix to the FormPatch (de1b0b2be0ed3efaade36e3783823d1b7a6af1ef) as a separate PR to master and 2.51.
Thank you

drewnoakes commented:
Pulled the patch fix out to #4781 

Added a few more fixes to this PR, so please review.

What do you mean by MVP?

RussKie commented:
> What do you mean by MVP?

MVP = Minimum viable product, in this case it is "minimum viable patch".

I don't think we'll have a stable v3 ready for release before July, so I want to release these fixes for 2.51 users.

drewnoakes commented:
@RussKie I'm relieved you're not asking me to build a new start-up :)

Happy to make another PR for the 2.51 branch. Will be less effort overall to do so once this one is considered ready to merge.

RussKie commented:
Over the weekend I'll sit in front of my big screen and check the proposed fix.

drewnoakes commented:
Sounds good. If you're in the mood to play with these changes a bit, search for controls that use `PictureBox` or buttons with images and put `this.AdjustForDpiScaling()` at the end of their constructors. That will scan the control tree and update the scaling factor for embedded images.

I've got a stash where I put this in a bunch of places but want to go through and review them properly before submitting.

```
$ git stash show --stat
 GitUI/CommandsDialogs/AboutBox.cs                  |  1 +
 .../BrowseDialog/DashboardControl/Dashboard.cs     |  2 ++
 .../BrowseDialog/FormCommitCount.cs                |  2 ++
 GitUI/CommandsDialogs/BrowseDialog/FormDonate.cs   |  2 ++
 GitUI/CommandsDialogs/FormAddToGitIgnore.cs        |  2 ++
 GitUI/CommandsDialogs/FormDeleteBranch.cs          |  1 +
 GitUI/CommandsDialogs/FormDeleteRemoteBranch.cs    |  1 +
 GitUI/CommandsDialogs/FormResolveConflicts.cs      |  1 +
 GitUI/CommandsDialogs/FormSettings.cs              |  1 +
 GitUI/CommandsDialogs/FormStash.cs                 |  1 +
 .../FormModifiedDeletedCreated.cs                  |  2 ++
 GitUI/CommandsDialogs/RevisionGpgInfo.cs           |  1 +
 .../SettingsDialog/Pages/ColorsSettingsPage.cs     |  1 +
 .../SettingsDialog/Pages/FormChooseTranslation.cs  |  1 +
 .../SettingsDialog/Pages/GitConfigSettingsPage.cs  |  1 +
 GitUI/Editor/FileViewer.cs                         |  1 +
 GitUI/FormPuttyError.cs                            |  1 +
 GitUI/FormStatus.cs                                |  3 +-
 GitUI/UserControls/GotoUserManualControl.cs        |  1 +
 GitUI/UserControls/HelpImageDisplayUserControl.cs  |  1 +
 GitUI/UserControls/RevisionGrid.cs                 |  1 +
 .../DeleteUnusedBranchesForm.cs                    |  2 ++
 Plugins/GitFlow/GitFlow.csproj                     |  4 +++
 Plugins/GitFlow/GitFlowForm.cs                     |  3 ++
```

I also was unable to come to a good resolution on tab control scaling in `FormBrowse` and `FormFileHistory`. It seems to work better on the latter, so at least it is possible.

Looking at this stuff last night made me miss WPF.

RussKie commented:
Nice!

I've seen recommendations to use Font scaling mode in a number of places, however we use Dpi scaling. @KindDragon do you remember why?

RussKie commented on GitExtUtils/GitUI/DpiUtil.cs:
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
```
Why has this moved from GitUI? This is UI specific...

drewnoakes replied:
On my phone. IIRC it was too make it available to a plugin.

RussKie replied:
ah, makes sense

drewnoakes commented:
Thanks for verifying. Even with the fuzzy icons it looks so much better at high DPI with these changes.

RussKie commented:
> I also was unable to come to a good resolution on tab control scaling in FormBrowse and FormFileHistory. 

The following seems to do the trick:

```diff
diff --git a/GitUI/CommandsDialogs/FormFileHistory.cs b/GitUI/CommandsDialogs/FormFileHistory.cs
index 24e8a450e..f7e6f37be 100644
--- a/GitUI/CommandsDialogs/FormFileHistory.cs
+++ b/GitUI/CommandsDialogs/FormFileHistory.cs
@@ -114,6 +114,8 @@ protected override void OnRuntimeLoad(EventArgs e)
         {
             base.OnRuntimeLoad(e);
 
+            tabControl1.ItemSize = DpiUtil.Scale(tabControl1.ItemSize);
+            tabControl1.Padding = DpiUtil.Scale(tabControl1.Padding);
+
             bool autoLoad = (tabControl1.SelectedTab == BlameTab && AppSettings.LoadBlameOnShow) || AppSettings.LoadFileHistoryOnShow;
```

![image](https://user-images.githubusercontent.com/4403806/38420373-da92da58-39e6-11e8-9d33-9d0b38036216.png)

I would also increase `tabControl1.Padding.Width` from 6 to 8 to make it slightly more visually pleasing.

RussKie commented:
The last 3 commits can probably be squashed without any loss of traceability

drewnoakes commented:
Was the changing of the tab header text removed? I don't see this on master any more, but it's there in 2.51. Setting the size explicitly might break this. If it's static, then that's fine, though we should do extra testing and maybe omit that change from a 2.51 patch.

![image](https://user-images.githubusercontent.com/350947/38427449-2e15628a-39b1-11e8-9207-5278df26ba8c.png)

RussKie commented:
Sorry Drrew you lost me with the question (it is well past midnight and I'm buggered).
Tab's padding is set to (6,3) IIRC, I proposed to make it (8,3) and scale with your helper to give headings a bit more chrome.
Tab's sizing is "normal" (default) - i.e. dynamic to accomodate the size of the text (I got it from MSDN), so the scaling should be safe.

drewnoakes commented:
Haha no worries. I wasn't very clear. Let me be more specific.

In the screenshot I posted, the diff tab's text says "Diff (A: parent -> B: selection)", but only when that tab is selected.

In `master` if you select the tab, the text remains as "Diff".

So I'm wondering if that's intentional, or a regression. I think I prefer the new behaviour anyway.

But if it's a regression and we fix it, then specifying the tab's size might not work. Speculation on my part.

I'll squash the last three commits.

drewnoakes commented:
Added some more fixes.

---

## FormAbout

![image](https://user-images.githubusercontent.com/350947/38440149-5217fb30-39d8-11e8-8338-425ad0d575fa.png)

![image](https://user-images.githubusercontent.com/350947/38440158-5a785f72-39d8-11e8-9a22-a72f59dab4d8.png)

---

## FormContributors

![image](https://user-images.githubusercontent.com/350947/38440250-c247379a-39d8-11e8-8652-67e4667e8f3c.png)

![image](https://user-images.githubusercontent.com/350947/38440254-c6a8d8a2-39d8-11e8-8fb9-8b0e2f070a2e.png)

---

## FormDonate

![image](https://user-images.githubusercontent.com/350947/38440284-de81fbac-39d8-11e8-8f69-8c867ed11f9e.png)

![image](https://user-images.githubusercontent.com/350947/38440294-e2bb0358-39d8-11e8-8a14-78b423390e54.png)

---
## FormDeleteBranch

![image](https://user-images.githubusercontent.com/350947/38440372-2c11de5a-39d9-11e8-8640-a86a9bfa6780.png)

![image](https://user-images.githubusercontent.com/350947/38440378-30af062c-39d9-11e8-8d88-e45561eb2ca9.png)

---

## RevisionGrid

![image](https://user-images.githubusercontent.com/350947/38441275-227f6b2a-39dc-11e8-8656-5f8cf01a1000.png)
![image](https://user-images.githubusercontent.com/350947/38441290-29db95b0-39dc-11e8-9ba0-a7a9f0eb3d1b.png)

---

## FormBrowse

![image](https://user-images.githubusercontent.com/350947/38441306-3bcce8b4-39dc-11e8-997f-097a5486e510.png)

![image](https://user-images.githubusercontent.com/350947/38441353-56b6bc2c-39dc-11e8-8f3b-826e49c7c90e.png)

drewnoakes commented:
@RussKie scaling the padding works well for tab controls 👍 

I added a clause to `AdjustForDpiScaling` that does that automatically.

Scaling `ItemSize` doesn't seem to work. The default value is (0,0), and any other value forces a given size on all tab headers, which we don't want. In a few places I removed explicit `ItemSize` and `Padding` properties from the designer generated code, as the default values work well. For the tab controls with images, those images have been scaled too.

Pretty happy with this PR now. Won't push anything else to it. Once it's merged I'll put together a PR for the 2.51 branch.

RussKie commented:
> Scaling ItemSize doesn't seem to work. The default value is (0,0), and any other value forces a given size on all tab headers, which we don't want. 

It is (0,0) until the form is loaded, hence in my example earlier I had it in `OnRuntimeLoad` method - this it works like a charm :)

RussKie commented:
@gerhardol could you please check https://github.com/gitextensions/gitextensions/pull/4772#issuecomment-379301882.
I think you worked in this area, are you able to comment? Thanks

drewnoakes commented:
> It is (0,0) until the form is loaded, hence in my example earlier I had it in OnRuntimeLoad method - this it works like a charm :)

My experiments with the ItemSize property showed it set all tabs to the same size, which isn't always what we want. If the passing approach works it'll simplify things as we can just call AdjustForDpiScaling from the constructor which solves many DPI problems in one go.

If you get a chance to test this out I'll be interested to hear your thoughts.

gerhardol commented:
@drewnoakes @RussKie 
> In master if you select the tab, the text remains as "Diff".

Yes, this is intentional, to avoid that the tabs resizes. Some users do not understand A-B anyway. When diff for "multiple A" was added, an explanation was needed anyway (there new some bugs with merge commits already too).

The affected revisions "first" and "Selected" can be seen in Reset and Diff.
(I opted to consistently use "First" also when only one revision was selected, that could have been Parent in those situations. Keeping the complexity down...)

It is possible, but I prefer to not merge the Diff tab change to 2.51, there were quite big changes, and there has been many related changes too.

RussKie commented on GitUI/CommandsDialogs/AboutBox.cs:
```diff
@@ -15,6 +15,7 @@ public AboutBox()
 
             InitializeComponent();
             Translate();
+            this.AdjustForDpiScaling();
```
why not do this on the base form/base usercontrol?

drewnoakes replied:
Can't think of a reason, so long as we test every form.

Should only be needed on the form, as it will find the user controls.

RussKie commented:
...unless user controls are added at runtime. there are few use cases

On 7 April 2018 at 10:51, Drew Noakes <notifications@github.com> wrote:

> *@drewnoakes* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/AboutBox.cs
> <https://github.com/gitextensions/gitextensions/pull/4772#discussion_r179902853>
> :
>
> > @@ -15,6 +15,7 @@ public AboutBox()
>
>              InitializeComponent();
>              Translate();
> +            this.AdjustForDpiScaling();
>
> Can't think of a reason, so long as we test every form.
>
> Should only be needed on the form, as it will find the user controls.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4772#discussion_r179902853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXqC6xPAOv8QGsWIddi-K4nfd0fB3ks5tmA2JgaJpZM4TG0Tq>
> .
>

RussKie commented:
On a second thought, I think my last comment is rubbish - I don't think we add custom usercontrols at runtime. And if we did, that would be developer's responsibility to call `AdjustForDpiScaling` on them.

I still think though, that `OnLoaded` or `OnRuntimeLoaded` are better place to perform scaling instead of the ctor.

RussKie commented:
I've checked the lot on W8 with 150% and it all looks good
